executor/index_lookup_merge_join.go

```diff
@@ -0,0 +1,673 @@
+// When err is not nil, task must not be nil to send the error to the main thread via task
+func (omw *outerMergeWorker) buildTask(ctx context.Context) (*lookUpMergeJoinTask, error) {
+	newFirstChunk(omw.executor)
+	task := &lookUpMergeJoinTask{
+		results:     make(chan *indexMergeJoinResult, 4),
+		doneErr:     make(chan error, 1),
+		outerResult: newFirstChunk(omw.executor),
+	}
+	task.memTracker = memory.NewTracker(stringutil.MemoizeStr(func() string { return fmt.Sprintf("lookup join task %p", task) }), -1)
+	task.memTracker.AttachTo(omw.parentMemTracker)
+
+	omw.increaseBatchSize()
+	if omw.lookup.isOuterJoin { // if is outerJoin, push the requiredRows down
+		requiredRows := int(atomic.LoadInt64(&omw.lookup.requiredRows))
+		task.outerResult.SetRequiredRows(requiredRows, omw.maxBatchSize)
+	} else {
+		task.outerResult.SetRequiredRows(omw.batchSize, omw.maxBatchSize)
+	}
+
+	task.memTracker.Consume(task.outerResult.MemoryUsage())
+	oldMemUsage := task.outerResult.MemoryUsage()
+	err := Next(ctx, omw.executor, task.outerResult)
+	if err != nil {
+		return task, err
+	}
+
+	newMemUsage := task.outerResult.MemoryUsage()
+	task.memTracker.Consume(newMemUsage - oldMemUsage)
+	if task.outerResult == nil || task.outerResult.NumRows() == 0 {
+		return nil, nil
+	}
+	if omw.outerMergeCtx.needOuterSort {
+		c := newFirstChunk(omw.executor)
+		var rows []chunk.Row
+		for i := 0; i < task.outerResult.NumRows(); i++ {
+			rows = append(rows, task.outerResult.GetRow(i))
+		}
+		sort.Slice(rows, func(i, j int) bool {
+			for id, joinKey := range omw.joinKeys {
+				cmp, _, err := omw.compareFuncs[id](omw.ctx, joinKey, joinKey, rows[i], rows[j])
+				terror.Log(err)
+				if cmp != 0 || omw.nextColCompareFilters == nil {
+					return cmp < 0
+				}
+				return omw.nextColCompareFilters.CompareRow(rows[i], rows[j]) < 0
+			}
+			return false
+		})
+		for i := 0; i < len(rows); i++ {
+			c.AppendRow(rows[i])
+		}
+		task.outerResult = c
+	}
+
+	if omw.filter != nil {
+		var err error
+		task.outerMatch, err = expression.VectorizedFilter(omw.ctx, omw.filter, chunk.NewIterator4Chunk(task.outerResult), task.outerMatch)
+		if err != nil {
+			return task, err
+		}
+		task.memTracker.Consume(int64(cap(task.outerMatch)))
+	}
+	task.outerIter = chunk.NewIterator4Chunk(task.outerResult)
+	return task, nil
+}
+
+func (omw *outerMergeWorker) increaseBatchSize() {
+	if omw.batchSize < omw.maxBatchSize {
+		omw.batchSize *= 2
+	}
+	if omw.batchSize > omw.maxBatchSize {
+		omw.batchSize = omw.maxBatchSize
+	}
+}
+
+func (imw *innerMergeWorker) run(ctx context.Context, wg *sync.WaitGroup, cancelFunc context.CancelFunc) {
+	var task *lookUpMergeJoinTask
+	defer func() {
+		wg.Done()
+		if r := recover(); r != nil {
+			cancelFunc()
+		}
+	}()
+
+	for ok := true; ok; {
+		select {
+		case task, ok = <-imw.taskCh:
+			if !ok {
+				return
+			}
+		case <-ctx.Done():
+			return
+		}
+
+		err := imw.handleTask(ctx, task)
+		task.doneErr <- err
```

Comment:
It may cause that the receiver receives `doneErr` before the `task.results` in `IndexLookUpMergeJoin.Next` and lose some data.

Reply:
Why? Only all the results have been sent to `results` chunk, the `err` sent to `doneErr` after.

Reply:
In `IndexLookUpMergeJoin.Next`, the reciever is 
```go
```

Reply:
But if both `results` channel and `doneErr` channel have the elements. It will get the result first, right?

Reply:
Unfortunately not.